README
bytes-circle v3.1

bytes-circle is a program that shows statistics about bytes 
contained in a file as an ASCII circle graph of deviations from 
mean in in standard deviation (sigma) fraction increments.

The program can be useful for statistically analyze the content of
files in a glimpse: text files are shown as a green centered crown,
compressed and encrypted files should be shown as equally distributed
variations with a very low CV (sigma/mean), and other types of files
can be classified between these two categories depending on their
ASCII vs binary content, which can be useful to quickly determine how
information is stored inside them (redis db, mmedia files, etc).


debian/changelog
bytes-circle (3.1-1) unstable; urgency=low

  * New upstream release.

 -- Roberto S. Galende <roberto.s.galende@gmail.com>  Mon,  2 Aug 2021 02:30:58 +0200

bytes-circle (3.0-1) unstable; urgency=low

  * New upstream release.

 -- Roberto S. Galende <roberto.s.galende@gmail.com>  Wed, 28 Jul 2021 20:24:58 +0200

bytes-circle (2.6-1) unstable; urgency=low

   * New upstream release.

 -- Roberto S. Galende <roberto.s.galende@gmail.com>  Thu, 06 Dec 2018 00:36:58 +0100

bytes-circle (2.5-1) unstable; urgency=low

  * New upstream release.

 -- Roberto S. Galende <roberto.s.galende@gmail.com>  Fri, 30 Dec 2016 19:31:08 +0100

bytes-circle (2.4-1) unstable; urgency=low

  * New upstream release.

 -- Roberto S. Galende <roberto.s.galende@gmail.com>  Mon, 05 Dec 2016 19:01:05 +0100

bytes-circle (2.3-1) unstable; urgency=low

  * Added use of stderr for error output.
  * stdin treatment recoded.
  * New command-line modifier -B

 -- Roberto S. Galende <roberto.s.galende@gmail.com>  Fri, 08 Jul 2016 19:45:33 +0200

bytes-circle (2.2-1) unstable; urgency=low

  * Initial release (Closes: #827555)

 -- Roberto S. Galende <roberto.s.galende@gmail.com>  Sun, 19 Jun 2016 11:45:16 +0200


src/circle.c
// Calculate statiscal variance of bytes in a file
// and graph them on an ASCII circle
//
// v1.0 by circulosmeos, 2015-10.
// v1.2 by circulosmeos, 2016-01.
// v2.1, v2.2 by circulosmeos, 2016-06.
// v2.3 by circulosmeos, 2016-07.
// v3.0, v3.1 by circulosmeos, 2021-07.
// wp.me/p2FmmK-96
// goo.gl/TNh5dq
//
// Distributed under GPL v3 or higher License.
//

#include "statistics_circle.h"

const signed int CIRCLE_EMPTY_VALUE=255;

const int MAX_X=35, MAX_Y=16;

const int MAX_VALUE=255;

// Stores in each byte bucket coordinates[MAX_VALUE + 1]
// the integer coordinates that correspond to it in the
// MAX_X * MAX_Y square containing the ASCII circle.
void create_circle(double complex *coordinates)
{

    //int MAX_VALUE=255;
    //int MAX_X=50, MAX_Y=16;
    int INIT_X=MAX_X/2, INIT_Y=MAX_Y/2;
    signed int circle[MAX_X][MAX_Y];
    //double complex coordinates[MAX_VALUE+1];
    //signed int CIRCLE_EMPTY_VALUE=255;
    double x ,y, r, inc_r, angle, inc_angle, proportion;
    int i, j, k;

    angle=0.0;
    inc_angle=5.0;
    r=1.0;
    inc_r=0.0184;
    proportion=0.5;
    x=INIT_X;
    y=INIT_Y;

    for (i=0; i<MAX_X; i++) {
        for (j=0; j<MAX_Y; j++) {
            circle[i][j]=CIRCLE_EMPTY_VALUE;
        }
    }

    circle[(int)x][(int)y]=0;
    coordinates[0] = (int)x + (int)y*I;
    x++;
    circle[(int)x][(int)y]=1;
    coordinates[1] = (int)x + (int)y*I;

    for (i=2; i<=MAX_VALUE; i++) {
        //printf ("%d=", i);
        do {
            k++;
            angle+=inc_angle;
            r+= inc_r;
            x+= -(cos(angle)) * r;
            y+= +(sin(angle)) * r * proportion;
            if (x>MAX_X || y>MAX_Y || x<0 || y<0) {
                fprintf (stderr, "ERROR! %.1f, %.1f", x, y );
                break;
            }
            if ( circle[(int)x][(int)y] == CIRCLE_EMPTY_VALUE ) {
                circle[(int)x][(int)y]=i;
            }
        } while (circle[(int)x][(int)y] != i);
        coordinates[i] = (int)x + (int)y*I;
    }

    //printf("\n%d",k);

}


src/statistics_circle.h
// Calculate statiscal variance of bytes in a file
// and graph them on an ASCII circle
//
// v1.0 by circulosmeos, 2015-10.
// v1.2 by circulosmeos, 2016-01.
// v2.1, v2.2 by circulosmeos, 2016-06.
// v2.3 by circulosmeos, 2016-07.
// v2.4 by circulosmeos, 2016-12.
// v3.0, v3.1 by circulosmeos, 2021-07.
// wp.me/p2FmmK-96
// goo.gl/TNh5dq
//
// Distributed under GPL v3 or higher License.
//

#ifndef STATISTICS_CIRCLE_H_
#define STATISTICS_CIRCLE_H_

#include <config.h>

#define _XOPEN_SOURCE 600 // getopt() compilation compatible with Solaris 9

// .................................................
// large file support (LFS) (files with size >2^31 (2 GiB) in linux, and >4 GiB in Windows)
#define _FILE_OFFSET_BITS 64    // stat, fseek
#define _LARGEFILE_SOURCE
#define _LARGEFILE64_SOURCE     // off64_t for fseek64
#define ZERO_LL 0LL             // crafted specially to be used in FSEEK( , , SEEK_END);
// .................................................


#include <stdio.h>
//#include <stdint.h> // incompatible with Solaris
#include <inttypes.h> // compatible with Solaris
#include <errno.h>    // printf to stderr
#include <complex.h>
#include <math.h>
#include <stdbool.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h> // getopt()
#include <ctype.h>  // isprint()
#include <sys/stat.h> //stat()

#ifndef __sun
# include <getopt.h> // getopt() compatible with -std=c99
#endif

// sets binary mode for stdin in Windows
#define STDIN 0
#ifdef _WIN32
# include <io.h>
# include <fcntl.h>
# define SET_BINARY_MODE(handle) setmode(handle, O_BINARY)
#else
# define SET_BINARY_MODE(handle) ((void)0)
#endif

#define BUFFER_LENGTH 4096

#define MAX_SIGMA_CHAR 10

#define KRED  "\x1B[31m"
#define KGRN  "\x1B[32m"
#define KBLU  "\x1B[34m"
#define KWHT  "\x1B[37m"
#define RESET "\x1B[m"

void create_circle( double complex *coordinates );

extern const signed int CIRCLE_EMPTY_VALUE;

extern const int MAX_X, MAX_Y;

extern const int MAX_VALUE;

void print_circle_value( signed int value );

int analyze_file(
    char *szFile,
    unsigned long long slice_number,
    unsigned long long slice_size,
    bool bShowGlobalFileStatistics,
    unsigned long long from_byte,
    unsigned long long to_byte
);

void empty_circle(
    signed int circle[MAX_X][MAX_Y],
    signed int circle2[MAX_X][MAX_Y],
    bool two_circles_flag
);

void calculate_sigma(
    unsigned long long *bytes,
    double *sigma_parameter,
    double *mean_parameter,
    int *number_of_byte_buckets_parameter,
    unsigned long long total_size
);

void print_circle_on_screen(
    unsigned long long *bytes,
    double sigma,
    double mean,
    double complex *coordinates,
    signed int circle[MAX_X][MAX_Y],
    signed int circle2[MAX_X][MAX_Y],
    bool two_circles_flag,
    int two_circles_value,
    bool restrict_statistics,
    int list_bytes,
    int number_of_byte_buckets,
    char *szFile,
    unsigned long long total_size,
    unsigned long long total_bytes_read,
    unsigned long long slice_size,
    unsigned long long slice_actual_number,
    unsigned long long from_byte,
    unsigned long long to_byte
);

void print_circle_value( signed int value );

void print_help();

unsigned long long giveMeAnInteger( const char *original_input );

#endif /* STATISTICS_CIRCLE_H_ */
